CMakeLists.txt

```diff
@@ -158,6 +159,48 @@ include(cmake/leveldb.cmake)
 include(cmake/minisketch.cmake)
 include(cmake/secp256k1.cmake)
 
+include(ProcessConfigurations)
+set_default_config(RelWithDebInfo)
+
+# Redefine configuration flags.
+add_compile_definitions($<$<CONFIG:Debug>:DEBUG>)
+add_compile_definitions($<$<CONFIG:Debug>:DEBUG_LOCKORDER>)
+add_compile_definitions($<$<CONFIG:Debug>:DEBUG_LOCKCONTENTION>)
+add_compile_definitions($<$<CONFIG:Debug>:RPC_DOC_CHECK>)
+add_compile_definitions($<$<CONFIG:Debug>:ABORT_ON_FAILED_ASSUME>)
+# We leave assertions on.
+if(MSVC)
+  remove_c_flag_from_all_configs(/DNDEBUG)
+  remove_cxx_flag_from_all_configs(/DNDEBUG)
+else()
+  remove_c_flag_from_all_configs(-DNDEBUG)
+  remove_cxx_flag_from_all_configs(-DNDEBUG)
+
+  # Prefer -O2 optimization level. (-O3 is CMake's default for Release for many compilers.)
+  replace_c_flag_in_config(Release -O3 -O2)
+  replace_cxx_flag_in_config(Release -O3 -O2)
+
+  set(debug_c_flags "")
+  set(debug_cxx_flags "")
+  try_append_cxx_flags(debug_cxx_flags "-O0" RESULT_VAR compiler_supports_O0)
+  if(compiler_supports_O0)
+    string(STRIP "${debug_c_flags} -O0" debug_c_flags)
+  endif()
+  try_append_cxx_flags(debug_cxx_flags "-g3" RESULT_VAR compiler_supports_g3)
+  if(compiler_supports_g3)
+    string(STRIP "${debug_c_flags} -g3" debug_c_flags)
+  else()
+    try_append_cxx_flags(debug_cxx_flags "-g")
+    string(STRIP "${debug_c_flags} -g" debug_c_flags)
+  endif()
+  try_append_cxx_flags(debug_cxx_flags "-ftrapv")
+  if(MINGW)
+    try_append_cxx_flags(debug_cxx_flags "-Wa,-mbig-obj")
```

Comment:
Where is this coming from?

Reply:
Otherwise, the debug build fails:
```
$ make -C depends HOST=x86_64-w64-mingw32 DEBUG=1 NO_QT=1
$ cmake -B build --toolchain depends/x86_64-w64-mingw32/share/toolchain.cmake -DCMAKE_BUILD_TYPE=Debug
$ cmake --build build --target bitcoin_wallet
...
[ 79%] Building CXX object src/wallet/CMakeFiles/bitcoin_wallet.dir/wallet.cpp.obj
In file included from /home/hebasto/git/bitcoin/src/wallet/wallet.cpp:6:
/home/hebasto/git/bitcoin/src/wallet/wallet.h: In function ‘std::string wallet::PurposeToString(wallet::AddressPurpose)’:
/home/hebasto/git/bitcoin/src/wallet/wallet.h:255:1: warning: control reaches end of non-void function [-Wreturn-type]
  255 | }
      | ^
/usr/bin/x86_64-w64-mingw32-as: CMakeFiles/bitcoin_wallet.dir/wallet.cpp.obj: too many sections (45678)
{standard input}: Assembler messages:
{standard input}: Fatal error: can't write 41 bytes to section .text of CMakeFiles/bitcoin_wallet.dir/wallet.cpp.obj: 'file too big'
/usr/bin/x86_64-w64-mingw32-as: CMakeFiles/bitcoin_wallet.dir/wallet.cpp.obj: too many sections (45678)
{standard input}: Fatal error: CMakeFiles/bitcoin_wallet.dir/wallet.cpp.obj: file too big
gmake[3]: *** [src/wallet/CMakeFiles/bitcoin_wallet.dir/build.make:437: src/wallet/CMakeFiles/bitcoin_wallet.dir/wallet.cpp.obj] Error 1
gmake[2]: *** [CMakeFiles/Makefile2:1555: src/wallet/CMakeFiles/bitcoin_wallet.dir/all] Error 2
gmake[1]: *** [CMakeFiles/Makefile2:1562: src/wallet/CMakeFiles/bitcoin_wallet.dir/rule] Error 2
gmake: *** [Makefile:732: bitcoin_wallet] Error 2
```

Reply:
Lol

Reply:
We should definitely look into breaking up that object. But in the meantime the obvious question... why don't we need this in master?

Reply:
> We should definitely look into breaking up that object. But in the meantime the obvious question... why don't we need this in master?

Good question :)

For Windows cross debug build:

- Autotools' flags are `-O0 -g3 -O1` that effectively means `-g3 -O1`
- CMake's flags are `-O1 -O0 -g3` that effectively means `-O0 -g3`

The behavior deviation needs to be fixed.